### PR TITLE
Use autoconf/automake/libtool to build the library

### PR DESCRIPTION
**Pull request by rajachan**
The configure.ac and Makefile.am configs are pretty rudimentary, but
they are better than the static Makefile that we have. Need to use
pkg-config to autodetect installation paths for the dependencies.

Signed-off-by: Raghu Raja <craghun@amazon.com>

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rajachan** [comment]
My autotools-fu is weak, but this should at least allow for flexible installs.

File: configure.ac [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+AC_ARG_WITH([mpi],
+            AC_HELP_STRING([--with-mpi=PATH], [Path to non-standard MPI installation]),
+            [AS_IF([test -d $withval/lib64], [mpi_libdir="lib64"], [mpi_libdir="lib"])
+             CFLAGS="-I$withval/include $CFLAGS"
+             LDFLAGS="-L$withval/$mpi_libdir $LDFLAGS"],
+            [])
+
+
+# Checks for header files.
+AC_CHECK_HEADERS([limits.h stdlib.h string.h unistd.h rdma/fabric.h cuda_runtime.h], [],[
```
[the remark this answers is not in the document]
Comment (rashikakheria) [reply configure.ac]: It didn't pick up the first time. Subsequent builds does. We can check a fresh checkout to verify.

[the remark this answers is not in the document]
Comment (rajachan) [reply configure.ac]: Tested a fresh checkout and this doesn't happen for me/

**Comment by rajachan** [comment]
I looks like the nccl-tests Makefile is not doing the right thing with non-standard NCCL installations and this should be fixed there instead of having to change the LD_LIBRARY_PATH at runtime:

```
$ ldd build/reduce_perf
	linux-vdso.so.1 =>  (0x00007ffc7dbc5000)
	libcudart.so.9.0 => /usr/local/cuda/lib64/libcudart.so.9.0 (0x00007f197a60b000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f197a403000)
	libcurand.so.9.0 => /usr/local/cuda/lib64/libcurand.so.9.0 (0x00007f197649f000)
	libnccl.so.2 => not found
	libnvToolsExt.so.1 => /usr/local/cuda/lib64/libnvToolsExt.so.1 (0x00007f1976296000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f197607a000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f1975e76000)
	libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f1975af1000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f19757ef000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f19755d9000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f197520c000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f197a878000)
```

File: configure.ac (hunk 1 above)
[the remark this answers is not in the document]
Comment (rashikakheria) [reply configure.ac]: Thanks for confirming!

**Pull request merged**